dlib/media/ffmpeg_abstract.h

```diff
@@ -0,0 +1,955 @@
+        };
+
+        // ---------------------------------------------------------------------------------------------------
+
+        struct codec_details
+        {
+            /*!
+                WHAT THIS OBJECT REPRESENTS
+                    This object informs on available codecs provided by the installation of ffmpeg dlib is linked against.
+            !*/
+            std::string codec_name;
+            bool supports_encoding;
+            bool supports_decoding;
+        };
+
+        struct device_details
+        {
+            /*!
+                WHAT THIS OBJECT REPRESENTS
+                    This object informs on available devices provided by the installation of ffmpeg dlib is linked against.
+            !*/
+
+            struct instance
+            {
+                std::string name;
+                std::string description;
+            };
+
+            std::string device_type;
+            std::vector<instance> devices;
+        };
+        
+        std::vector<std::string> list_protocols();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg protocols
+        !*/
+
+        std::vector<std::string> list_demuxers();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg demuxers
+        !*/
+
+        std::vector<std::string> list_muxers();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg muxers
+        !*/
+        
+        std::vector<codec_details> list_codecs();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg codecs with information on whether decoding and/or encoding is supported.
+                  Note that not all codecs support encoding, unless your installation of ffmpeg is built with third party library
+                  dependencies like libx264, libx265, etc.
+        !*/
+
+        std::vector<device_details> list_input_devices();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg input devices and available instances of those devices
+        !*/
+
+        std::vector<device_details> list_output_devices();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg output devices and available instances of those devices
+        !*/
+
+        // ---------------------------------------------------------------------------------------------------
+
+        template <class ImageContainer>
+        void convert(const frame& f, ImageContainer& image)
+        /*!
+            requires
+                - f.is_image() == true
+                - f.pixfmt() == pix_traits<typename image_traits<ImageContainer>::pixel_type>::fmt
+
+            ensures
+                - converts a frame object into array2d<rgb_pixel>
+        !*/
+
+        template <class ImageContainer>
+        void convert(const ImageContainer& img, frame& f)
+        /*!
+            ensures
```

Comment:
Same thing here.  Add `requires` saying what the image type must be.  Add that anywhere there is a template argument.  The `requires` part needs to say what the type is supposed to be.  At some point we will have concepts and that will make it nicer, but not today :(